### PR TITLE
Use platform runtime check for ProtectedData

### PR DESCRIPTION
**Pull request by vcsjones**
This changes S.S.C.ProtectedData to use runtime checks instead of TPI to determine if that platform is Windows or not.

This fixes an issue where `ProtectedData` does not work on UWP in Windows because UWP is picking up the .NET Standard version. We [removed](https://github.com/dotnet/runtime/pull/64610) .NET Standard 2.0 platform specific target for Windows, which previously allowed this to work in UWP. Since NS2.0-windows is no longer valid, this moves ProtectedData to use a runtime check for all implementations.

Since .NET Standard is being supported again, some refactoring that have since occurred need to be undone, or accounted for, now that we are no longer using `GeneratePlatformNotSupportedAssemblyMessage` for NS2.0. Some of the Common/* sources have been refactored to work off of `Span<T>`. So for .NETStandard we take a dependency on `System.Memory`. We can avoid this if we want to create non-span local copies of those common helpers. I have not deeply investigated how feasible this is.

Lastly, this adds some unit tests to make sure we are correctly throwing PNSE on non-Windows.

Contributes to #78875

### CODE REVIEW

**Review by bartonjs — APPROVED** [review]
Looks a lot like a change I started before I hibernated through all of December :smile:

**Review by ViktorHofer — APPROVED** [review]
Thanks a lot 👍

File: src/libraries/System.Security.Cryptography.ProtectedData/src/System/Security/Cryptography/ProtectedData.cs [hunk 1]
```diff
@@ -16,14 +16,20 @@ public static partial class ProtectedData
 
         public static byte[] Protect(byte[] userData, byte[]? optionalEntropy, DataProtectionScope scope)
```
Comment (marek-safar) [inline src/libraries/System.Security.Cryptography.ProtectedData/src/System/Security/Cryptography/ProtectedData.cs:17]: It's missing SupportedOSPlatformAttribute annotation

File: src/libraries/System.Security.Cryptography.ProtectedData/src/System/Security/Cryptography/ProtectedData.cs [hunk 2]
```diff
@@ -16,14 +16,20 @@ public static partial class ProtectedData
...
-            ArgumentNullException.ThrowIfNull(userData);
+            CheckPlatformSupport();
+
+            if (userData is null)
+                throw new ArgumentNullException(nameof(userData));
 
             return ProtectOrUnprotect(userData, optionalEntropy, scope, protect: true);
         }
 
         public static byte[] Unprotect(byte[] encryptedData, byte[]? optionalEntropy, DataProtectionScope scope)
```
Comment (marek-safar) [inline src/libraries/System.Security.Cryptography.ProtectedData/src/System/Security/Cryptography/ProtectedData.cs:27]: same here

File: src/libraries/System.Security.Cryptography.ProtectedData/src/System/Security/Cryptography/ProtectedData.cs (hunk 1 above)
Comment (ViktorHofer) [reply src/libraries/System.Security.Cryptography.ProtectedData/src/System/Security/Cryptography/ProtectedData.cs:17]: Isn't this sufficient? https://github.com/vcsjones/runtime/blob/3d98834cd3cf6d0ce017c74cdd1dd32bacd0d5de/src/libraries/System.Security.Cryptography.ProtectedData/Directory.Build.props#L5

Comment (vcsjones) [reply src/libraries/System.Security.Cryptography.ProtectedData/src/System/Security/Cryptography/ProtectedData.cs:17]: When I look at the library in the locally built package, I see the `SupportedOSPlatformAttribute("windows")` on the whole assembly:

```
  .custom instance void [System.Runtime]System.Runtime.Versioning.SupportedOSPlatformAttribute::.ctor(string) = ( 01 00 07 77 69 6E 64 6F 77 73 00 00 )             // ...windows..
```

So think we're fine.

Comment (marek-safar) [reply src/libraries/System.Security.Cryptography.ProtectedData/src/System/Security/Cryptography/ProtectedData.cs:17]: missed that, so we should be fine.

**Pull request merged**